File: strategy/marvin/group/striker.lua
local Striker = Class("Group.Striker")

local Field = require "../base/field"
local vis = require "../base/vis"
local World = require "../base/world"
local G = World.Geometry
local MovesHelper = require "util/moveshelper"

function Striker:init()
	self.name = "striker"

	self._strikerCount = 0

	self._zones = {}
	self._emptyZone = nil

	self._lastMainAttacker = nil
end

local function getDefaultPosition(boundaries)
	local zoneWidth = boundaries.right - boundaries.left
	local zoneHeight = boundaries.top - boundaries.bottom
	local x, y
	repeat
		x = (math.random() * 0.6 + 0.2) * zoneWidth + boundaries.left
		y = (math.random() * 0.6 + 0.2) * zoneHeight + boundaries.bottom
	until not Field.isInOpponentDefenseArea(Vector(x, y), 0.2)
	return Vector(x, y)
end

local function visualizeZone(zone)
	local edge = 0.05
	local left = zone.boundaries.left + edge
	local right = zone.boundaries.right - edge
	local top = zone.boundaries.top - edge
	local bottom = zone.boundaries.bottom + edge
	local points = { Vector(left, top), Vector(left, bottom), Vector(right, bottom), Vector(right, top), Vector(left, top) }
	vis.addPath("g/striker: Zones", points, vis.colors.gold, nil, nil, 0.02)
end

local function assignRobotsToZones(robotPositions, zones)
	local n = #zones
	if n == 0 then
		return {}
	end

	local positions = {}
	local robots = {}
	for robot, robotPos in pairs(robotPositions) do
		table.insert(positions, {pos = robotPos})
		table.insert(robots, robot)
	end
	local zonePositions = {}
	for _, zone in ipairs(zones) do
		table.insert(zonePositions, zone.defaultPos)
	end
	local assignment = MovesHelper.assignRobots(positions, zonePositions, 0)

	local zoneAssignment = {}
	for i, zone in ipairs(zones) do
		zoneAssignment[zone] = robots[assignment[i]]
	end
	return zoneAssignment
end

function Striker:_updateZones(robots)
	local totalLeft = -G.FieldWidthHalf
	local totalRight = G.FieldWidthHalf
	local totalTop = G.FieldHeightHalf
	local totalBottom = -G.FieldHeightQuarter

	local nStrikers = #robots
	local remainingZones = nStrikers + 1 -- one zone will stay empty
	self._strikerCount = nStrikers

	-- reset the zones
	self._zones = {}
	if remainingZones == 0 then
		return
	end

	-- create midfield zone
	do
		local boundaries = { left = totalLeft, right = totalRight, top = G.FieldHeightHalf/4, bottom = totalBottom }
		local defaultPos = getDefaultPosition(boundaries)
		table.insert(self._zones, {boundaries = boundaries, defaultPos = defaultPos})
		remainingZones = remainingZones - 1
	end

	-- create offensive zones
	local zoneWidth = (totalRight - totalLeft) / remainingZones
	for i = 1, remainingZones do
		local boundaries = { left = totalLeft + (i - 1) * zoneWidth, right = totalLeft + i * zoneWidth,
				top = totalTop, bottom = G.FieldHeightHalf / 4 }
		local defaultPos = getDefaultPosition(boundaries)
		table.insert(self._zones, {boundaries = boundaries, defaultPos = defaultPos})
	end

	-- reset empty zone hysteresis
	self._emptyZone = nil
end

function Striker:_chooseEmptyZone(mainAttackerPos)
	local emptyZoneHysteresis = self._emptyZone and 0.2 or 0
	if mainAttackerPos then
		for _, zone in ipairs(self._zones) do
			if mainAttackerPos.x >= zone.boundaries.left + emptyZoneHysteresis
					and mainAttackerPos.x <= zone.boundaries.right - emptyZoneHysteresis 
					and mainAttackerPos.y >= zone.boundaries.bottom + emptyZoneHysteresis
					and mainAttackerPos.y <= zone.boundaries.top - emptyZoneHysteresis then
				self._emptyZone = zone
				break
			end
		end
	end

	-- default: midfield zone is empty
	if not self._emptyZone and #self._zones > 0 then
		self._emptyZone = self._zones[1]
	end
end

function Striker:run(sender, inbox, messages)
	local robots = table.keys(messages)
	local mainAttacker = inbox.mainAttacker().trainer

	-- if the mainAttacker changes, assume that the previous mainAttacker becomes a striker instead
	local robotsTmp = {}
	for _, robot in ipairs(robots) do
		if robot == mainAttacker and self._lastMainAttacker then
			table.insert(robotsTmp, self._lastMainAttacker)
		else
			table.insert(robotsTmp, robot)
		end
	end
	robots = robotsTmp

	-- update zones if necessary
	if #robots ~= self._strikerCount then
		self:_updateZones(robots)
	end

	-- choose which zone is occupied by the mainAttacker
	local mainAttackerPos = nil
	if mainAttacker then
		mainAttackerPos = inbox.attackPosition()[mainAttacker] or mainAttacker.pos
	end
	self:_chooseEmptyZone(mainAttackerPos)

	-- assign the zones to the nearest strikers
	local robotPositions = {} -- robot -> pos
	local _, passInfoTable = next(inbox.passInfo())
	for _, r in ipairs(robots) do
		local pos = r.pos
		if passInfoTable then
			for _, passInfo in ipairs(passInfoTable) do
				if passInfo.target == r then
					pos = passInfo.ballPos + (passInfo.ballPos - World.Ball.pos):setLength(r.shootRadius + World.Ball.radius)
				end
			end
		end
		robotPositions[r] = pos
	end

	local zoneList = {} -- { zone }
	for _, zone in ipairs(self._zones) do
		if zone ~= self._emptyZone then
			table.insert(zoneList, zone)
			visualizeZone(zone)
		end
	end

	local robotZones = assignRobotsToZones(robotPositions, zoneList)

	for zone, robot in pairs(robotZones) do
		sender.strikerZone(robot, zone)
	end

	self._lastMainAttacker = mainAttacker
end

return Striker

File: strategy/typescript/glados/group/striker.ts
import * as Field from "base/field";
import * as MathUtil from "base/mathutil";
import { FriendlyRobot } from "base/robot";
import { Position, Vector } from "base/vector";
import * as vis from "base/vis";
import * as World from "base/world";

import { MessageBox, MessageType } from "glados/control/messaging";
import { Group } from "glados/trainer/groups";
import * as MovesHelper from "glados/util/moveshelper";

let G = World.Geometry;

interface Boundaries {
	right: number;
	left: number;
	top: number;
	bottom: number;
}

interface Zone {
	boundaries: Boundaries;
	defaultPos: Position;
}

function getDefaultPosition(boundaries: Boundaries): Position {
	const zoneWidth = boundaries.right - boundaries.left;
	const zoneHeight = boundaries.top - boundaries.bottom;
	let x, y;
	do {
		x = (MathUtil.random() * 0.6 + 0.2) * zoneWidth + boundaries.left;
		y = (MathUtil.random() * 0.6 + 0.2) * zoneHeight + boundaries.bottom;
	} while (Field.isInOpponentDefenseArea(new Vector(x, y), 0.2));
	return new Vector(x, y);
}

function visualizeZone(zone: Zone) {
	let edge = 0.05;
	let left = zone.boundaries.left + edge;
	let right = zone.boundaries.right - edge;
	let top = zone.boundaries.top - edge;
	let bottom = zone.boundaries.bottom + edge;
	let points = [ new Vector(left, top), new Vector(left, bottom), new Vector(right, bottom), new Vector(right, top) ];
	vis.addPolygon("g/striker: Zones", points, vis.colors.gold);
}

function assignRobotsToZones(robotPositions: Map<FriendlyRobot, Position>, zones: Zone[]): Map<Zone, FriendlyRobot> {
	let n = zones.length;
	if (n === 0) {
		return new Map<Zone, FriendlyRobot>();
	}

	let positions: {pos: Position}[] = [];
	let robots: FriendlyRobot[] = [];
	for (let [robot, robotPos] of robotPositions.entries()) {
		positions.push({pos: robotPos});
		robots.push(robot);
	}
	let zonePositions: Position[] = [];
	for (let zone of zones) {
		zonePositions.push(zone.defaultPos);
	}
	let assignment = MovesHelper.assignRobots(positions, zonePositions);

	let zoneAssignment: Map<Zone, FriendlyRobot> = new Map<Zone, FriendlyRobot>();
	for (let i = 0;i < zones.length;i++) {
		const zone = zones[i];
		zoneAssignment[zone] = robots[assignment[i]];
	}

	// visualize assignments
	if (!amun.isPerformanceMode) {
		for (let [zone, robot] of zoneAssignment.entries()) {
			vis.addPath("g/striker: zone assignment", [zone.defaultPos, robot.pos], vis.colors.white);
		}
	}
	return zoneAssignment;
}

export class Striker implements Group {
	readonly name = "striker";

	_strikerCount: number = 0;

	_zones: Zone[] = [];
	_emptyZone: Zone | undefined;

	_lastMainAttacker: FriendlyRobot | undefined;
	_lastRobots: FriendlyRobot[] | undefined;
	_lastAssignments: Map<Zone, FriendlyRobot> | undefined = undefined;


	_updateZones(robots: FriendlyRobot[]) {
		let totalLeft = -G.FieldWidthHalf;
		let totalRight = G.FieldWidthHalf;
		let totalTop = G.FieldHeightHalf;
		let totalBottom = -G.FieldHeightQuarter;

		let nStrikers = robots.length;
		let remainingZones = nStrikers + 1; // one zone will stay empty
		this._strikerCount = nStrikers;

		// reset the zones
		this._zones = [];
		if (remainingZones === 0) {
			return;
		}

		// create midfield zone
		{
			let boundaries = { left: totalLeft, right: totalRight, top: G.FieldHeightHalf / 4, bottom: totalBottom };
			let defaultPos = getDefaultPosition(boundaries);
			this._zones.push({boundaries: boundaries, defaultPos: defaultPos});
			remainingZones = remainingZones - 1;
		}

		// create offensive zones
		let zoneWidth = (totalRight - totalLeft) / remainingZones;
		for (let i = 1;i <= remainingZones;i++) {
			let boundaries = { left: totalLeft + (i - 1) * zoneWidth, right: totalLeft + i * zoneWidth,
					top: totalTop, bottom: G.FieldHeightHalf / 4 };
			let defaultPos = getDefaultPosition(boundaries);
			this._zones.push({boundaries: boundaries, defaultPos: defaultPos});
		}

		// reset empty zone hysteresis
		this._emptyZone = undefined;
	}

	_chooseEmptyZone(mainAttackerPos?: Position) {
		let emptyZoneHysteresis = this._emptyZone ? 0.2 : 0;
		if (mainAttackerPos != undefined) {
			for (let zone of this._zones) {
				if (mainAttackerPos.x >= zone.boundaries.left + emptyZoneHysteresis
						&&  mainAttackerPos.x <= zone.boundaries.right - emptyZoneHysteresis
						&&  mainAttackerPos.y >= zone.boundaries.bottom + emptyZoneHysteresis
						&&  mainAttackerPos.y <= zone.boundaries.top - emptyZoneHysteresis) {
					this._emptyZone = zone;
					break;
				}
			}
		}

		// default: midfield zone is empty
		if (!this._emptyZone && this._zones.length > 0) {
			this._emptyZone = this._zones[0];
		}
	}

	run(messaging: MessageBox, messages: Map<FriendlyRobot, any>) {
		let robots = Array.from(messages.keys());
		let mainAttacker = messaging.receiveTrainer(MessageType.mainAttacker);
		let prevEmptyZone = this._emptyZone;

		// if the mainAttacker changes, assume that the previous mainAttacker becomes a striker instead
		let robotsTmp: FriendlyRobot[] = [];
		for (let robot of robots) {
			if (robot === mainAttacker && this._lastMainAttacker) {
				robotsTmp.push(this._lastMainAttacker);
			} else {
				robotsTmp.push(robot);
			}
		}
		robots = robotsTmp;

		// update assignments if necessary
		let updateAssignments = this._lastRobots == undefined || this._lastAssignments == undefined || robots.length !== this._lastRobots.length;
		if (!updateAssignments) {
			for (let i = 0;i < robots.length;i++) {
				if (robots[i] !== (this._lastRobots as FriendlyRobot[])[i]) {
					updateAssignments = true;
					break;
				}
			}
		}

		// update zones if necessary
		if (robots.length !== this._strikerCount) {
			updateAssignments = true;
			this._updateZones(robots);
		}

		// choose which zone is occupied by the mainAttacker
		let mainAttackerPos = undefined;
		if (mainAttacker) {
			mainAttackerPos = messaging.receiveSingleSender(MessageType.attackPosition)[1] || mainAttacker.pos;
		}
		this._chooseEmptyZone(mainAttackerPos);

		// update assignments if empty zone changed
		updateAssignments = updateAssignments || this._emptyZone !== prevEmptyZone;

		// assign the zones to the nearest strikers
		let robotPositions = new Map<FriendlyRobot, Position>(); // robot -> pos
		let passInfoTable = messaging.receiveSingleSender(MessageType.passInfo)[1];
		for (let r of robots) {
			let pos = r.pos;
			if (passInfoTable) {
				for (let passInfo of passInfoTable) {
					if (passInfo.target === r) {
						pos = passInfo.ballPos + (passInfo.ballPos - World.Ball.pos).setLength(r.shootRadius + World.Ball.radius);
					}
				}
			}
			robotPositions[r] = pos;
		}

		let zoneList: Zone[] = []; // { zone }
		for (let zone of this._zones) {
			if (zone !== this._emptyZone) {
				zoneList.push(zone);
				visualizeZone(zone);
			}
		}

		let robotZones = updateAssignments ? assignRobotsToZones(robotPositions, zoneList) : <Map<Zone, FriendlyRobot>> this._lastAssignments;

		for (let [zone, robot] of robotZones.entries()) {
			messaging.send(MessageType.strikerZone, robot, zone);
		}


		this._lastMainAttacker = mainAttacker;
		this._lastRobots = robots;
		this._lastAssignments = robotZones;
	}
}
